src/main/java/com/example/bookstoreapp/service/impl/OrderServiceImpl.java

```diff
@@ -0,0 +1,131 @@
+    private final OrderMapper mapper;
+    private final ShoppingCartService shoppingCartService;
+    private final OrderItemMapper orderItemMapper;
+    private final OrderItemServiceImpl orderItemService;
+    private final BookRepository bookRepository;
+    private final UserService userService;
+
+    @Override
+    protected JpaRepository<Order, Long> getRepository() {
+        return repository;
+    }
+
+    @Override
+    protected OrderMapper getMapper() {
+        return mapper;
+    }
+
+    @Override
+    public OrderResponseDto placeOrder(PlaceOrderRequestDto placeOrderDto) {
+        ShoppingCart shoppingCart = shoppingCartService.findCartByUserId();
+
+        Order order = mapper.toOrderFromCart(shoppingCart);
+        order.setShippingAddress(placeOrderDto.getShippingAddress());
+        order.setStatus(Order.Status.COMPLETED);
+        Order savedOrder = repository.save(order);
+
+        Set<OrderItem> orderItems = getOrderItemsFromCart(shoppingCart);
+        orderItems.forEach(orderItem -> orderItem.setOrder(savedOrder));
+        savedOrder.setOrderItems(getSavedOrderItems(orderItems));
+        completePurchase(shoppingCart);
+        return mapper.toDto(savedOrder);
+    }
+
+    @Override
+    public List<OrderResponseDto> getOrderHistory(Pageable pageable) {
+        return repository.getAllByUser(pageable, userService.getUser())
+                .stream()
+                .map(mapper::toDto)
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public List<OrderItemResponseDto> getOrderItems(Long orderId) {
+        return getOrderById(orderId)
+                .getOrderItems()
+                .stream()
+                .map(orderItemMapper::toDto)
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public void update(Long orderId, UpdateStatusRequestDto updateStatusRequestDto) {
+        Order orderFromDb = getOrderById(orderId);
+        orderFromDb.setStatus(updateStatusRequestDto.getStatus());
+        repository.save(orderFromDb);
+    }
+
+    private Set<OrderItem> getOrderItemsFromCart(ShoppingCart shoppingCart) {
+
+        return shoppingCart.getCartItems()
+                .stream()
+                .map(cartItem -> orderItemMapper.cartItemToOrderItem(cartItem,
+                        getBookFromCartItem(cartItem)))
+                .collect(Collectors.toSet());
+    }
+
+    private Book getBookFromCartItem(CartItem cartItem) {
+        return bookRepository.findById(cartItem.getBook().getId()).orElseThrow(() ->
+                new EntityNotFoundException("Can't find book by id: "
+                        + cartItem.getBook().getId()));
+    }
+
+    private Set<OrderItem> getSavedOrderItems(Set<OrderItem> orderItems) {
+        return orderItems
+                .stream()
+                .map(orderItemService::save)
+                .collect(Collectors.toSet());
+    }
+
+    public void completePurchase(ShoppingCart shoppingCart) {
```

Comment:
If the method is used only inside the class, it must be private
```suggestion
    private void completePurchase(ShoppingCart shoppingCart) {
```